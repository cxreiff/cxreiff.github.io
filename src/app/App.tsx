import React from 'react'
import { Provider } from 'react-redux'
import { ConnectedRouter } from 'connected-react-router'

import 'normalize.css'

import store from './store'
import history from './history'

import Layout from '~/src/features/layout/layout.component'

const App: React.FC = () => (
    <Provider store={store}>
        <ConnectedRouter history={history}>
            <Layout/>
        </ConnectedRouter>
    </Provider>
)

export default App
