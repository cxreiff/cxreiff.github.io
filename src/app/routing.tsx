import { FC, ReactNode, Suspense, useEffect, lazy } from "react";
import { Routes, Route, useLocation } from "react-router-dom";

import { Loader } from "src/common/loader/loader.component";
import { IFrame } from "src/common/iframe/iframe.component";
import {
  DOCUMENT_TITLE_ROOT,
  BREAKOUT_URL,
  PLATFORMER_URL,
  HEX_TILES_URL,
  PIXIJS_URL,
  BABYLONJS_URL,
  PLAIN_WEBGL_URL,
  BIGFOOT_MAP_URL,
} from "src/utilities/constants";

import Homepage from "src/features/homepage/homepage.component";
import Projects from "src/features/projects/projects.component";
import Photos from "src/features/photos/photos.component";
import Sketches from "src/features/sketches/sketches.component";
import ProjectsFigrid from "src/features/projects/projectsFigrid.component";

// const Posts = lazy(() => import("src/features/posts/posts.component"));
// const PostsContent = lazy(
//   () => import("src/features/posts/postsContent.component")
// );

const Shaders = lazy(() => import("src/features/shaders/shaders.component"));

const Pixijs = () => (
  <IFrame aspect={15.5 / 9} url={PIXIJS_URL} title={"pixijs"} canFullscreen />
);
const Babylonjs = () => (
  <IFrame
    aspect={3 / 2}
    url={BABYLONJS_URL}
    title={"babylonjs"}
    canFullscreen
  />
);
const Breakout = () => (
  <IFrame
    aspect={3 / 2}
    mobileAspect={2 / 3}
    url={BREAKOUT_URL}
    title={"breakout"}
    canFullscreen
  />
);
const Platformer = () => (
  <IFrame
    aspect={16 / 10}
    url={PLATFORMER_URL}
    title={"platformer"}
    canFullscreen
  />
);
const HexTiles = () => (
  <IFrame
    aspect={16 / 10}
    url={HEX_TILES_URL}
    title={"hex tiles"}
    canFullscreen
  />
);
const PlainWebGL = () => (
  <IFrame
    aspect={16 / 10}
    url={PLAIN_WEBGL_URL}
    title={"hex tiles"}
    canFullscreen
  />
);
const BigfootMap = () => (
  <IFrame aspect={16 / 10} url={BIGFOOT_MAP_URL} title={"bigfoot sightings"} />
);

export type RouteDefinition = {
  path: string;
  Component: FC;
  children?: RouteDefinition[];
};

export const ROUTES: RouteDefinition[] = [
  { path: "", Component: Homepage },
  {
    path: "projects",
    Component: Projects,
    children: [
      {
        path: "figrid",
        Component: ProjectsFigrid,
      },
      {
        path: "pixijs",
        Component: Pixijs,
        children: [{ path: ":id", Component: Pixijs }],
      },
      { path: "babylonjs", Component: Babylonjs },
      { path: "breakout", Component: Breakout },
      { path: "platformer", Component: Platformer },
      { path: "hex_tiles", Component: HexTiles },
      { path: "webgl", Component: PlainWebGL },
      { path: "bigfoot", Component: BigfootMap },
      {
        path: "shaders",
        Component: Shaders,
        children: [{ path: ":category", Component: Shaders }],
      },
      {
        path: "sketches",
        Component: Sketches,
        children: [{ path: ":id", Component: Sketches }],
      },
    ],
  },
  { path: "photos", Component: Photos },
  // {
  //   path: "posts",
  //   Component: Posts,
  //   children: [{ path: ":id", Component: PostsContent }],
  // },
];

export const Routing = () => {
  const { pathname } = useLocation();

  useEffect(() => {
    const baseRoute = pathname.split("/")[1];
    document.title = baseRoute
      ? `${DOCUMENT_TITLE_ROOT} â€” ${baseRoute}`
      : DOCUMENT_TITLE_ROOT;
  }, [pathname]);

  const renderRoutes = (routes: RouteDefinition[], prefix = "") =>
    routes.reduce((renderedRoutes, { path, Component, children }) => {
      const fullPath = `${prefix}/${path}`;

      if (children) {
        renderedRoutes.push(...renderRoutes(children, fullPath));
      }

      return renderedRoutes.concat(
        <Route path={fullPath} element={<Component />} />
      );
    }, [] as ReactNode[]);

  return (
    <Suspense fallback={<Loader />}>
      <Routes>{renderRoutes(ROUTES)}</Routes>
    </Suspense>
  );
};
