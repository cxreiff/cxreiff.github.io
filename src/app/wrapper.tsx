import React, { ComponentType, ReactElement } from 'react'
import { Provider } from 'react-redux'
import { ConnectedRouter } from 'connected-react-router'

import store from './store'
import history from './history'

const wrapper = (Component: ComponentType, props = {}): ReactElement => {
  return (
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <Component {...props} />
      </ConnectedRouter>
    </Provider>
  )
}

export default wrapper
