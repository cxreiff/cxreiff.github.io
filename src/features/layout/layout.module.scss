@use '../../shared.scss' as *;
@use '../../themed.scss' as *;

.layout {
    max-width: $md-breakpoint;
    margin: auto;

    &>main {
        margin: 0 1rem 2rem;
    }

    &>nav {
        position: relative;
        margin: 2rem 1rem;
        
        @include box_shadow();

        @include themed(true) {
            background-color: t($color-primary);
        }

        ul {
            position: relative;
            z-index: 0;
            display: flex;
            flex-direction: column;
            margin: 0;
            padding: 0;
            list-style: none;
            min-height: 3rem;

            @include is_larger_than($sm-breakpoint) {
                flex-direction: row;
                margin-right: 3.1rem;
            }
        }

        li {
            height: 3rem;
            width: 100%;

            a {
                position: relative;
                display: inline-block;
                text-align: center;
                cursor: pointer;
                text-decoration: none;
                font-size: 1.3rem;
                height: 100%;
                width: 100%;
                vertical-align: middle;
                line-height: 3rem;
                letter-spacing: -0.02em;
                
                @include themed(true) {
                    color: t($color-darker);
                }
            }
        }

        li.logo {
            z-index: 1;
            flex: 1;
            width: calc(100% - 6rem);
            margin: 0 3rem 1.7rem 3rem;

            @include is_larger_than($sm-breakpoint) {
                width: 100%;
                margin: 0;
            }

            a {
                width: 100%;
                margin-top: -0.4rem;
                white-space: nowrap;
                font-size: 4.8rem;
                letter-spacing: -0.05em;

                @include is_larger_than($sm-breakpoint) {
                    width: min-content;
                    padding: 0 1rem;
                
                    &:before {
                        content: '{';
                        padding-right: 1rem;
                    }
                    &:after {
                        content: '}';
                        padding-left: 1rem;
                    }
                    &:before, &:after {
                        font-size: 2rem;
                        transform: translateY(-12px);
                        display: inline-block;
                        transition: transform 0.1s ease-out, text-shadow 0.1s ease-out;
                        
                        @include themed(true) {
                            color: t($color-tertiary);
                        }
                    }
                    @media (hover: hover) {
                        &:hover:before, &:hover:after {
                            transform: translateY(-12px) scale(1.1);
                        }
                        &:hover:before {
                            @include themed(true) {
                                text-shadow: -1px 0px 0 t($color-tertiary);
                            }
                        }
                        &:hover:after {
                            @include themed(true) {
                                text-shadow: 1px 0px 0 t($color-tertiary);
                            }
                        }
                    }
                }
            }
        }

        li.navlink {
            position: relative;
            flex: 1;

            @include is_larger_than($sm-breakpoint) {
                flex: 0 1 140px;
            }

            &>a {
                @include interactive(
                    0.3rem,
                    $roundness,
                    $color-dark,
                    $color-darker,
                    $color-darker,
                    $color-lighter,
                );
            }
        }

        li.current {
            a {
                text-decoration: underline;
                
                @include themed(true) {
                    color: t($color-lighter);
                    background: t($color-dark);
                }
            }
        }
            
        @include is_larger_than($sm-breakpoint) {
            li.navlink:not(.current):after {
                content: '';
                opacity: 0.2;
                position: absolute;
                height: 60%;
                width: 1px;
                top: 20%;
                right: -2px;
                
                @include themed(true) {
                    border-left: 2px solid t($color-dark);
                }
            }
        }
        @include is_smaller_than($sm-breakpoint) {
            li.navlink:not(.current):not(:first-child):not(:last-child):after {
                content: '';
                opacity: 0.2;
                position: absolute;
                width: 50%;
                height: 1px;
                left: 25%;
                bottom: -2px;
                
                @include themed(true) {
                    border-top: 2px solid t($color-dark);
                }
            }
        }

        /* Hiding and showing navbar at mobile sizes. */
        ul {
            max-height: 3rem;

            @include is_larger_than($sm-breakpoint) {
                max-height: inherit;
            }
        }
        li.navlink {
            display: none;

            @include is_larger_than($sm-breakpoint) {
                display: inherit;
            }
        }
        &.expanded {
            ul {
                max-height: inherit;
            }

            li.navlink {
                display: inherit;
            }
        }
    }
}

.menu_button {
    position: absolute;
    z-index: 1;
    top: 0;
    left: 0;

    @include is_larger_than($sm-breakpoint) {
        display: none;
    }
}
