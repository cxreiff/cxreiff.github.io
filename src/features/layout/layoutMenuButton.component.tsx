import { FC } from 'react'
import { Menu } from 'lucide-preact'

import { IconButton } from '~/src/common/iconButton/iconButton.component'

import styles from './layoutMenuButton.module.scss'

type LayoutMenuButtonProps = {
    toggleMenu: () => void,
}

export const LayoutMenuButton: FC<LayoutMenuButtonProps> = ({ toggleMenu }) => (
    <IconButton
        aria-label={'expand navigation'}
        className={styles.menu_button}
        onClick={toggleMenu}
    >
        <Menu />
    </IconButton>
)
