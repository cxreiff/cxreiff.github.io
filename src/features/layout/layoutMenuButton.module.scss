@use '../../shared.scss' as *;

.menu_button {
    position: absolute;
    z-index: 1;
    top: 0;
    left: 0;

    @include is_larger_than($sm-breakpoint) {
        display: none;
    }
}
