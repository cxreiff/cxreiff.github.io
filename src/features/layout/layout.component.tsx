import { FC, ReactNode, useState, useLayoutEffect } from "react";
import { useNavigate, useLocation } from "react-router-dom";
import { Menu } from "lucide-preact";
import cn from "classnames";

import { RouteDefinition } from "~/src/app/routing";
import { IconButton } from "~/src/common/iconButton/iconButton.component";
import { ThemeToggle } from "~/src/features/theme/themeToggle.component";

import styles from "./layout.module.scss";

type LayoutProps = {
  routes: RouteDefinition[];
  children?: ReactNode;
};

export const Layout: FC<LayoutProps> = ({ children, routes }) => {
  const pathname = useLocation().pathname;
  const navigate = useNavigate();
  const [expanded, setExpanded] = useState(false);

  useLayoutEffect(() => setExpanded(false), [pathname]);

  const comparePaths = (path: string, pathname: string) =>
    path === pathname.split("/")[1];

  return (
    <div className={styles.layout}>
      <nav
        aria-label={"primary"}
        className={cn({ [styles.expanded]: expanded })}
      >
        <IconButton
          aria-label={"expand navigation"}
          className={styles.menu_button}
          onClick={() => setExpanded(!expanded)}
        >
          <Menu />
        </IconButton>
        <ul>
          <li aria-label={"navigate to home"} className={styles.logo}>
            <a onClick={() => navigate("")}>cooper reiff</a>
          </li>
          {routes
            .filter(({ path }) => path)
            .map(({ path }) => (
              <li
                key={path}
                aria-label={`navigate to ${path}`}
                className={cn(styles.navlink, {
                  [styles.current]: comparePaths(path, pathname),
                })}
              >
                <a onClick={() => navigate(path)}>{path}</a>
              </li>
            ))}
        </ul>
        <ThemeToggle className={styles.theme_toggle} />
      </nav>
      <main>{children}</main>
    </div>
  );
};
