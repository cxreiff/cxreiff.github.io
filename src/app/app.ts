import { FC } from 'react'
import 'normalize.css'

import wrapper from './wrapper'
import Layout from '~/src/features/layout/layout.component'

const App: FC = () => wrapper(Layout)

export default App
