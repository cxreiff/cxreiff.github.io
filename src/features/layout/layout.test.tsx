import { render, fireEvent } from '@testing-library/preact'

import { Wrapper } from '~/src/app/wrapper'
import { ROUTES } from '~/src/app/routing'
import { Layout } from './layout.component'

const mockNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))

test('should render', () => {
    const { container } = render(<Layout routes={ROUTES} />, { wrapper: Wrapper })

    expect(container).toMatchSnapshot()
})

test('should expand menu', () => {
    const { container, getByRole } = render(<Layout routes={ROUTES} />, { wrapper: Wrapper })

    fireEvent.click(getByRole('button', { name: 'expand navigation' }))

    expect(container).toMatchSnapshot()
})

test('should navigate via navigation menu links', () => {
    const { getByRole } = render(<Layout routes={ROUTES} />, { wrapper: Wrapper })

    fireEvent.click(getByRole('listitem', { name: 'navigate to home' }).firstElementChild!)
    expect(mockNavigate).toHaveBeenCalledWith('')

    fireEvent.click(getByRole('listitem', { name: 'navigate to projects' }).firstElementChild!)
    expect(mockNavigate).toHaveBeenCalledWith('projects')

    fireEvent.click(getByRole('listitem', { name: 'navigate to photos' }).firstElementChild!)
    expect(mockNavigate).toHaveBeenCalledWith('photos')

    fireEvent.click(getByRole('listitem', { name: 'navigate to posts' }).firstElementChild!)
    expect(mockNavigate).toHaveBeenCalledWith('posts')
})
