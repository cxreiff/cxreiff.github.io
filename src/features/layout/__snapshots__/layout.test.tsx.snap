// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout should navigate to photos tab 1`] = `
<Layout
  routes={
    Array [
      Object {
        "Component": [Function],
        "children": Array [
          Object {
            "Component": [Function],
            "path": "pixijs",
          },
          Object {
            "Component": [Function],
            "path": "babylonjs",
          },
        ],
        "path": "projects",
      },
      Object {
        "Component": [Function],
        "path": "photos",
      },
      Object {
        "Component": [Function],
        "children": Array [
          Object {
            "Component": [Function],
            "path": ":id",
          },
        ],
        "path": "posts",
      },
      Object {
        "Component": [Function],
        "path": "resume",
      },
    ]
  }
>
  <div
    className="layout"
  >
    <nav
      aria-label="primary"
    >
      <ul>
        <li
          className="logo"
        >
          <a
            onClick={[Function]}
          >
            jax reiff
          </a>
        </li>
        <li
          className="navlink"
          id="nav-projects"
          key="projects"
        >
          <a
            onClick={[Function]}
          >
            projects
          </a>
        </li>
        <li
          className="navlink current"
          id="nav-photos"
          key="photos"
        >
          <a
            onClick={[Function]}
          >
            photos
          </a>
        </li>
        <li
          className="navlink"
          id="nav-posts"
          key="posts"
        >
          <a
            onClick={[Function]}
          >
            posts
          </a>
        </li>
        <li
          className="navlink"
          id="nav-resume"
          key="resume"
        >
          <a
            onClick={[Function]}
          >
            resume
          </a>
        </li>
      </ul>
    </nav>
    <main />
  </div>
</Layout>
`;

exports[`layout should navigate to posts tab 1`] = `
<Layout
  routes={
    Array [
      Object {
        "Component": [Function],
        "children": Array [
          Object {
            "Component": [Function],
            "path": "pixijs",
          },
          Object {
            "Component": [Function],
            "path": "babylonjs",
          },
        ],
        "path": "projects",
      },
      Object {
        "Component": [Function],
        "path": "photos",
      },
      Object {
        "Component": [Function],
        "children": Array [
          Object {
            "Component": [Function],
            "path": ":id",
          },
        ],
        "path": "posts",
      },
      Object {
        "Component": [Function],
        "path": "resume",
      },
    ]
  }
>
  <div
    className="layout"
  >
    <nav
      aria-label="primary"
    >
      <ul>
        <li
          className="logo"
        >
          <a
            onClick={[Function]}
          >
            jax reiff
          </a>
        </li>
        <li
          className="navlink"
          id="nav-projects"
          key="projects"
        >
          <a
            onClick={[Function]}
          >
            projects
          </a>
        </li>
        <li
          className="navlink"
          id="nav-photos"
          key="photos"
        >
          <a
            onClick={[Function]}
          >
            photos
          </a>
        </li>
        <li
          className="navlink current"
          id="nav-posts"
          key="posts"
        >
          <a
            onClick={[Function]}
          >
            posts
          </a>
        </li>
        <li
          className="navlink"
          id="nav-resume"
          key="resume"
        >
          <a
            onClick={[Function]}
          >
            resume
          </a>
        </li>
      </ul>
    </nav>
    <main />
  </div>
</Layout>
`;

exports[`layout should navigate to projects tab 1`] = `
<Layout
  routes={
    Array [
      Object {
        "Component": [Function],
        "children": Array [
          Object {
            "Component": [Function],
            "path": "pixijs",
          },
          Object {
            "Component": [Function],
            "path": "babylonjs",
          },
        ],
        "path": "projects",
      },
      Object {
        "Component": [Function],
        "path": "photos",
      },
      Object {
        "Component": [Function],
        "children": Array [
          Object {
            "Component": [Function],
            "path": ":id",
          },
        ],
        "path": "posts",
      },
      Object {
        "Component": [Function],
        "path": "resume",
      },
    ]
  }
>
  <div
    className="layout"
  >
    <nav
      aria-label="primary"
    >
      <ul>
        <li
          className="logo"
        >
          <a
            onClick={[Function]}
          >
            jax reiff
          </a>
        </li>
        <li
          className="navlink current"
          id="nav-projects"
          key="projects"
        >
          <a
            onClick={[Function]}
          >
            projects
          </a>
        </li>
        <li
          className="navlink"
          id="nav-photos"
          key="photos"
        >
          <a
            onClick={[Function]}
          >
            photos
          </a>
        </li>
        <li
          className="navlink"
          id="nav-posts"
          key="posts"
        >
          <a
            onClick={[Function]}
          >
            posts
          </a>
        </li>
        <li
          className="navlink"
          id="nav-resume"
          key="resume"
        >
          <a
            onClick={[Function]}
          >
            resume
          </a>
        </li>
      </ul>
    </nav>
    <main />
  </div>
</Layout>
`;

exports[`layout should navigate to resume tab 1`] = `
<Layout
  routes={
    Array [
      Object {
        "Component": [Function],
        "children": Array [
          Object {
            "Component": [Function],
            "path": "pixijs",
          },
          Object {
            "Component": [Function],
            "path": "babylonjs",
          },
        ],
        "path": "projects",
      },
      Object {
        "Component": [Function],
        "path": "photos",
      },
      Object {
        "Component": [Function],
        "children": Array [
          Object {
            "Component": [Function],
            "path": ":id",
          },
        ],
        "path": "posts",
      },
      Object {
        "Component": [Function],
        "path": "resume",
      },
    ]
  }
>
  <div
    className="layout"
  >
    <nav
      aria-label="primary"
    >
      <ul>
        <li
          className="logo"
        >
          <a
            onClick={[Function]}
          >
            jax reiff
          </a>
        </li>
        <li
          className="navlink"
          id="nav-projects"
          key="projects"
        >
          <a
            onClick={[Function]}
          >
            projects
          </a>
        </li>
        <li
          className="navlink"
          id="nav-photos"
          key="photos"
        >
          <a
            onClick={[Function]}
          >
            photos
          </a>
        </li>
        <li
          className="navlink"
          id="nav-posts"
          key="posts"
        >
          <a
            onClick={[Function]}
          >
            posts
          </a>
        </li>
        <li
          className="navlink current"
          id="nav-resume"
          key="resume"
        >
          <a
            onClick={[Function]}
          >
            resume
          </a>
        </li>
      </ul>
    </nav>
    <main />
  </div>
</Layout>
`;

exports[`layout should render 1`] = `
<Layout
  routes={
    Array [
      Object {
        "Component": [Function],
        "children": Array [
          Object {
            "Component": [Function],
            "path": "pixijs",
          },
          Object {
            "Component": [Function],
            "path": "babylonjs",
          },
        ],
        "path": "projects",
      },
      Object {
        "Component": [Function],
        "path": "photos",
      },
      Object {
        "Component": [Function],
        "children": Array [
          Object {
            "Component": [Function],
            "path": ":id",
          },
        ],
        "path": "posts",
      },
      Object {
        "Component": [Function],
        "path": "resume",
      },
    ]
  }
>
  <div
    className="layout"
  >
    <nav
      aria-label="primary"
    >
      <ul>
        <li
          className="logo"
        >
          <a
            onClick={[Function]}
          >
            jax reiff
          </a>
        </li>
        <li
          className="navlink"
          id="nav-projects"
          key="projects"
        >
          <a
            onClick={[Function]}
          >
            projects
          </a>
        </li>
        <li
          className="navlink"
          id="nav-photos"
          key="photos"
        >
          <a
            onClick={[Function]}
          >
            photos
          </a>
        </li>
        <li
          className="navlink"
          id="nav-posts"
          key="posts"
        >
          <a
            onClick={[Function]}
          >
            posts
          </a>
        </li>
        <li
          className="navlink"
          id="nav-resume"
          key="resume"
        >
          <a
            onClick={[Function]}
          >
            resume
          </a>
        </li>
      </ul>
    </nav>
    <main />
  </div>
</Layout>
`;
